### PR TITLE
Feature/wall

### PR DESCRIPTION
**Pull request by sockperson**
address #215, adding support for Mercurial Gauntlet

- adds a table to store player Mercurial Gauntlet quest progress
- player Mercurial Gauntlet progress is imported upon importing a save file

the following have not been implemented yet:

- any Mercurial Gauntlet endeavours
- rupies/mana dropping from the MG boss after it is defeated in a quest (this is visual)
- actually granting the monthly MG reward list (logic for this is done-- just don't know where you would grant the rewards)

### CODE REVIEW

**Comment by SapiensAnatis** [comment]
Closes #215

File: DragaliaAPI/Features/Wall/WallController.cs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        this.rewardService = rewardService;
+        this.clearPartyService = clearPartyService;
+        this.dungeonService = dungeonService;
+        this.wallService = wallService;
+        this.logger = logger;
+    }
+
+    // Called when failing a MG quest
+    // should return the ID and level of the failed quest in the response...
+    // but not sure how to get that from the dungeon key in the request currently
```
Comment (SapiensAnatis) [inline DragaliaAPI/Features/Wall/WallController.cs:42]: maybe would be best to use the sessions then -- can write something for MG

File: DragaliaAPI/Features/Wall/WallController.cs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+                    quest_id = session.QuestData.Id,
+                    wall_id = 0,
+                    wall_level = 0,
+                    is_host = true,
+                }
+            }
+        );
+    }
+
+    // When is this called?
```
Comment (SapiensAnatis) [inline DragaliaAPI/Features/Wall/WallController.cs:65]: checking monthly reward maybe?

File: DragaliaAPI/Features/Dungeon/QuestEnemyService.cs [hunk 3]
```diff
@@ -207,4 +216,45 @@ private IEnumerable<AtgenEnemy> GetEnemyList(int questId, int areaNum)
...
+        }
+
+        if (!enemies.Enemies.TryGetValue(VariationTypes.Normal, out IEnumerable<int>? enemyList))
+        {
+            this.logger.LogWarning(
+                "Unable to retrieve enemy list for variation type {type}",
+                VariationTypes.Normal
+            );
+            return Enumerable.Empty<AtgenEnemy>();
+        }
```
Comment (SapiensAnatis) [inline DragaliaAPI/Features/Dungeon/QuestEnemyService.cs:246]: Is any of this actually needed if you just use `questWallDetail.BossEnemyParamId`

File: DragaliaAPI/Features/Dungeon/QuestEnemyService.cs [hunk 4]
```diff
@@ -207,4 +216,45 @@ private IEnumerable<AtgenEnemy> GetEnemyList(int questId, int areaNum)
...
+
+        return enemyList.Select(
```
Comment (SapiensAnatis) [inline DragaliaAPI/Features/Dungeon/QuestEnemyService.cs:248]: doesn't need to be a select for just 1 element

File: DragaliaAPI/Features/SavefileUpdate/V13Update.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
+﻿using DragaliaAPI.Database.Repositories;
+using DragaliaAPI.Features.Wall;
+using DragaliaAPI.Services.Game;
+
+namespace DragaliaAPI.Features.SavefileUpdate;
+
+public class V13Update(IWallRepository wallRepository, IStoryRepository storyRepository)
+    : ISavefileUpdate
+{
+    public int SavefileVersion => 13;
```
Comment (SapiensAnatis) [inline DragaliaAPI/Features/SavefileUpdate/V13Update.cs:10]: This should be changed to v15

File: DragaliaAPI/Features/Wall/WallRepository.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+                    WallLevel = 0,
+                    IsStartNextLevel = false,
+                }
+            );
+        }
+    }
+
+    public async Task<bool> IsInitialized()
+    {
+        return await this.QuestWalls.CountAsync() > 0;
```
Comment (SapiensAnatis) [inline DragaliaAPI/Features/Wall/WallRepository.cs:58]:
```suggestion
        return await this.QuestWalls.AnyAsync()
```

**Review by SapiensAnatis — CHANGES_REQUESTED** [review]

File: DragaliaAPI/Features/Dungeon/QuestEnemyService.cs (hunk 3 above)
Comment (sockperson) [reply DragaliaAPI/Features/Dungeon/QuestEnemyService.cs:246]: ah when I wrote the code originally I just took it from the already existing enemy generating code, but it should just be as simple as generating a list of 1 enemy with the proper boss param id. Fixed in a recent commit

**Review by SapiensAnatis — APPROVED** [review]
The day of the `wall` is finally upon us

**Pull request merged**